src/System.Collections/src/System/Collections/Generic/Stack.cs

```diff
@@ -262,6 +262,7 @@ public bool TryPop(out T result)
         }
 
         // Pushes an item to the top of the stack.
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```

Comment:
> But we shouldn't add AggressiveInlining unless there's a strong performance-motivated scenario for doing so. Getting something to inline isn't a goal in and of itself.

(source: https://github.com/dotnet/corefx/pull/12094#issuecomment-249834575)

For the microbenchmark this may be good, but as the call-sites get bigger, it may decrease perf in real world scenarios.

I don't know what the choice is...

Reply:
I.e.

<details>
  <summary>this would be "inlined" into every call-site</summary>

```asm
G_M60461_IG02:
       8B4714               mov      eax, dword ptr [rdi+20]
       8B5718               mov      edx, dword ptr [rdi+24]
       488B4F08             mov      rcx, gword ptr [rdi+8]
       448B4108             mov      r8d, dword ptr [rcx+8]
       443BC2               cmp      r8d, edx
       7623                 jbe      SHORT G_M60461_IG04
       443BC0               cmp      r8d, eax
       761E                 jbe      SHORT G_M60461_IG04
       4C63C8               movsxd   r9, eax
       4289748910           mov      dword ptr [rcx+4*r9+16], esi
       FFC0                 inc      eax
       443BC0               cmp      r8d, eax
       7502                 jne      SHORT G_M60461_IG03
       33C0                 xor      eax, eax

G_M60461_IG03:
       894714               mov      dword ptr [rdi+20], eax
       FFC2                 inc      edx
       895718               mov      dword ptr [rdi+24], edx
       FF471C               inc      dword ptr [rdi+28]
       EB05                 jmp      SHORT G_M60461_IG05

G_M60461_IG04:
       E810F5FFFF           call     MyQueue`1:EnqueueWithResize(int):this
``` 
</details>

(" " because the JIT may optimize something, depending on the call-site).

Reply:
This is right. @stephentoub @jkotas when deciding where aggressive is reasonable, do you have rules of thumb? Eg, known to be heavily called in hot paths + known to typically fail automatic inlining + not very large?

Reply:
Ups, this dasm / comment belongs to Queue.

Reply:
+ known to improve perf significantly

Reply:
> deciding where aggressive is reasonable, do you have rules of thumb?

My rule of thumb is if you're using it, you most likely shouldn't be. I generally only use it when it's a critical-path function, when inlining makes a very measurable impact on consumers (not just a microbenchmark of the method itself), and there's something about it that would fool the JIT's heuristics, such as a ton of IL but most of which will be shaken away from the resulting asm, e.g. lots of different branches for different generic type tests against various struct types, intrinsics usage, etc. I don't think this qualifies. I'm actually skeptical of the existing usage on List.Add.

Reply:
Right, every use of AggressiveInlining should come with numbers (not just trivial micro-benchmarks) to justify its use.

Reply:
This aspect of performance optimization is interesting, is there somewhere some documentation/research about it(explain issue on "real system" out of microbenchmarks) or is only "experience"?

Reply:
Inlining a method results in a copy of its code at the call site.  With lots of call sites, that can significantly bloat code size.  That increase in code size is not just an issue in and of itself (e.g. more memory consumption, etc.), but each copy of that code is fundamentally different, recognized as different code for an instruction cache; having many different copies of the same code can thus result in the instruction cache being much less efficient.

Reply:
I've removed the use of `AggressiveInlining` on these. That guidance is good, thanks @stephentoub - I was purely running off what had been applied to List here.

Reply:
@stephentoub thank's for explanation, I wonder if there are "numbers" somewhere(maybe on codebase or on "internet") or if we should take it as "best practice", similar to "remove null check "here" is unuseful because branch predictor likely guesses", I'm aware that is very slippery slope but maybe you(and internal team researcher) have some real number

Reply:
This is highly dependent on actual code and actual call-sites. What numbers do you expect?
Research wise this optimization domain is covered, you may find a lot of papers, but the decision if inline or not comes back to actual code. So there won't be any hard rules, just heuristics.